### ISSUE TITLE
Context override + bpy.ops.object.delete conflicting with depsgraph handler

### ISSUE THREAD

**Issue by tonpix** [semagnum/light-painter#69]
1. put in monkey mesh
2. go Render view
3. select Mesh Light Paint
4. click once on mesh
5. press Escape
6. blender crashes (every time)

Blender 4.1.1

Example clip (did more clicks than it's needed to force crash)
https://files.catbox.moe/9f22jq.mp4

**Comment by semagnum** [comment]
Thanks for the repro steps and clip! Really appreciate that, makes things easier to debug.

While it doesn't crash for me, your video and my side confirm something else likely causing it: the monkey mesh stays selected and active while using the tool. You can see the monkey mesh in your example suddenly become a convex hull right before crashing. And since the tool expects to work on the _new_ mesh only, it starts operating on the monkey head, mesh data goes crazy, and for you (and likely others), crashes.

I'll fix that selection issue (over the weekend, sooner than later) but double check here if that resolves the issue.

**Comment by tonpix** [comment]
I have bad news, I managed to get consistent crash while mesh isn't selected - used Light Paint, clicked few times, then pressed Escape key and blender crashed 0.5s later. Not sure if crash is still related to selection issue, or no, but throwing another clip:

https://files.catbox.moe/rps4mo.mp4

**Comment by tonpix** [comment]
I also tested the same torus scenario with deselecting all before activating Paint Light tool - result was same

**Comment by tonpix** [comment]
Sorry for spam, but wanted to report I had no issues in 4.2, using addon installed through Extension platform and from disc - no crashes at all. Guess it nails problem down to blender 4.1.1

**Comment by semagnum** [comment]
Gotcha. Well, I'm currently on 4.1.1 and cannot reproduce the crash for any of the steps you gave :/ I'll try downloading from extensions platform to see if the code is the culprit for some reason (it shouldn't be).

In the meantime, could you check if previous Light Painter releases have the same problem or if it's only on the most recent one? That could narrow it down to a code change.

**Comment by tonpix** [comment]
Tried 1.3.0 and 1.2.8, but both were crashing. Is it possible to launch blender temporarily clean to see if it wasn't caused by conflict with other addons/settings?

**Comment by semagnum** [comment]
> Is it possible to launch blender temporarily clean to see if it wasn't caused by conflict with other addons/settings?

In the executable folder, there should be a script to run with factory settings. On Windows it's `C:\Program Files\Blender Foundation\Blender 4.1\blender_factory_startup.cmd`, otherwise you can use the `--factory-startup` in the terminal.

**Comment by tonpix** [comment]
I've used --factory-startup argument just now to launch blender 4.1.1 clean, no crashes it seems. Guess some addon/setting is conflicting, Gonna be bit hard to narrow it down :( I'll check it further after some sleep

**Comment by tonpix** [comment]
Nevermind, it went faster and found the culprit - it's Adobe Substance 3D addon. It can be downloaded from https://www.adobe.com/products/substance3d/plugins/substance-in-blender.html# (also requires another 3rd party files from same link - Integration tool) - in case you'd like to investigate what's causing crashing while pressing Esc key.

![blender_q1xLWFBfvh](https://github.com/semagnum/light-painter/assets/13473971/d6ba1bde-7a53-44fe-8a89-65be54744d16)

**Comment by semagnum** [comment]
Tested downloading the Substance add-on, could reproduce the crash. Will close the ticket then as it's a third-party add-on conflict. I'll still make the ticket regarding the mesh tool still selecting the active object when it shouldn't.

I'll speculate, just from glancing at Substance's code: it's their handler function called every time the depsgraph (ie the scene and its geometry) changes:

```py
# in Substance add-on's __init__.py
bpy.app.handlers.depsgraph_update_post.append(SUBSTANCE_Persistance.depsgraph_update_post)
```

The depsgraph handler checks selected objects and sets some scene variables. But Blender can easily get stuck calling these functions over and over again if it unintentionally changes the scene, prompting that change calling the handler _again_, and so on. A potentially infinite loop happens, Blender crashes to save itself. I try to avoid them for that reason 😄 and they can slow down viewport performance if it's not coded right.

If someone finds a code solution for Light Painter that resolves it (without breaking cancel/cleanup behavior), I'll happily merge it. But until then, I have a feeling it's on their end.

**Comment by semagnum** [comment]
Hold on - on a hunch, I tried a way of removing the newly made light a different way, and Substance didn't complain. I'll give something a try this weekend and update here.

**Comment by semagnum** [comment]
Alright, I pushed a change that stops the crashes with the Substance add-on enabled. Can you confirm that for me please before I make a release? Just download as a zip

**Comment by tonpix** [comment]
Looks it doesn't crash. Great job 👍 Thanks for quick fix. 

But sometimes I was getting error message when pressing Enter key while using Mesh Light Paint, but I don't know how to repro it yet (pretty much I had to click around like a monkey) and this issue doesn't seem to be affecting functionality - light polygon is still created. Error message doesn't happen with Light Paint. For now I can only throw clip, if I figure out over when error message happens, I'll let you know.

https://files.catbox.moe/lmm546.mp4

**Comment by semagnum** [comment]
I can replicate it on my own side - looks to be an issue with re-selecting mesh objects prior to using the tool. That change was to fix #70.

I'll reopen that ticket and close this one.
